### PR TITLE
JAMES-3142 API Segragation: UnitializedEventBus // EventBus

### PR DESCRIPTION
**Pull request by chibenwa**
Points to work on:

We have dependencies loops: some listeners requires either  managers that themselves requires an event bus that requires the listeners for its initialisation.

Guice handling of this dependency loop is straightforward.

However for unit tests this is tricky to deal with and I experimented several strategies:

 - The **evil** MagicEventBus** interface holding both initialized & uninitialized EventBus

This simplifies test setup a lot and avoids sistematic wrapping.

Then the unitialiazeed event bus and the event bus are the same object, we can pass the unitialized event bus, then later initialize it.

This limits the refactoring benefits however calls to MagicEventBus api can be tracked. We can rely on the API split where we need (eg in guice) while skipping it where it hurts (unit tests + spring). However theeventBus implementation cannot be split.

 - Spring compatibility strategy relies on inVm event but to be implementing both

 - the alternative to the evil **MagicEventBus** would be systematic wrapping  of object taking a eventBus to kill dependencies loop. In short: manual re-implementation of Guice dependency loop killing. That is what I implemented in "InMemoryIntegrationResources".

A third alternative could be to rely on guice for initializing these test systems.

-------------------------------------

Follow up to this work:

 - this is not simple to split `EventBus` and `UninitializedEventBus` APIs.
     - Thus we might want to accept https://github.com/linagora/james-project/pull/3279 as an intermediate implementation stage
     - We need to discuss which one of the three strategies to adopt above.
     - Note that these strategies could be composed: eg start with the magic API solution (which makes segragation effective in production code) then enact the wrapping/guice-for-testing strategy

Also we might want to measure the payoff of such a refactoring over https://github.com/linagora/james-project/pull/327

As there is no silver bullet this need discussion!

### CODE REVIEW

**Comment by Arsnael** [comment]
```
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project apache-james-mailbox-cassandra: Compilation failure: Compilation failure: 
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR] /james-project/mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraMessageIdManagerTestSystem.java:[56,37] cannot find symbol
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR]   symbol:   method initialize(org.apache.james.mailbox.store.quota.ListeningCurrentQuotaUpdater)
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR]   location: interface org.apache.james.mailbox.events.EventBus
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR] /james-project/mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraMessageIdManagerSideEffectTest.java:[33,1] org.apache.james.mailbox.cassandra.CassandraMessageIdManagerSideEffectTest is not abstract and does not override abstract method createTestSystem(org.apache.james.mailbox.quota.QuotaManager,org.apache.james.mailbox.events.UninitializedEventBus,java.util.Set<org.apache.james.mailbox.extension.PreDeletionHook>) in org.apache.james.mailbox.store.AbstractMessageIdManagerSideEffectTest
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR] /james-project/mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraMessageIdManagerSideEffectTest.java:[38,5] method does not override or implement a method from a supertype
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR] -> [Help 1]
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project apache-james-mailbox-jpa: Compilation failure
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR] /james-project/mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/JpaMailboxManagerStressTest.java:[49,74] method provideMailboxManager in class org.apache.james.mailbox.jpa.JpaMailboxManagerProvider cannot be applied to given types;
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR]   required: org.apache.james.backends.jpa.JpaTestCluster,com.google.common.collect.ImmutableList<org.apache.james.mailbox.events.MailboxListener.GroupMailboxListener>
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR]   found: org.apache.james.backends.jpa.JpaTestCluster
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR]   reason: actual and formal argument lists differ in length
[084adf61cd51fb67c344fcd0b687435d8d0d4a03] [ERROR] -> [Help 1]
```

File: mailbox/api/src/main/java/org/apache/james/mailbox/events/EventBus.java [hunk 1]
```diff
@@ -26,6 +26,9 @@
 import reactor.core.publisher.Mono;
 
 public interface EventBus {
+
+    interface MagicEventBus extends EventBus, UninitializedEventBus {}
```
Comment (Arsnael) [inline mailbox/api/src/main/java/org/apache/james/mailbox/events/EventBus.java:30]: Should we really call something `Magic`?

Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/events/EventBus.java:30]: No.

Read the PR description.

**Comment by mbaechler** [comment]
What about injecting an (memoized) EventBus supplier into Managers to break the loop?

uninitializedEventBus

(if you want to edit the diagram, here is the source http://www.plantuml.com/plantuml/uml/TP5Foi8m4CNtESLSe3r05n_b4mL1mOKU82L37OY9pYHTADxT_ccWq7GvUMz-xv6f9MWEyM7LpceHE944TCZpU7FaA92sz4Qpxz25_oZmKG1D96igEAvbzPsMSzgi7Bm_QwTRP3M_9CE5VKS6USnuJiDMYafbmHV5Wba1uoiIewWLjYp_ShGgPObAd0cxAhRKjs6niIJtgqyLHs_-DuUsxU3YtfvrREUCH-ZNQdIcl-4F )

**Comment by chibenwa** [comment]
Gonna give it a try.

**Comment by chibenwa** [comment]
Maildir mpt tests

**Comment by chibenwa** [comment]
```
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithCassandraBlobStoreTest.mailsContentWithUnicodeCharactersShouldBeKeptUnChanged » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithCassandraBlobStoreTest.mailsShouldBeWellReceived » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithCassandraBlobStoreTest.oneHundredMailsShouldBeWellReceived » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithDefaultAwsS3Test.mailsContentWithUnicodeCharactersShouldBeKeptUnChanged » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithDefaultAwsS3Test.mailsShouldBeWellReceived » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithDefaultAwsS3Test.oneHundredMailsShouldBeWellReceived » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithDefaultSwiftTest.mailsContentWithUnicodeCharactersShouldBeKeptUnChanged » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithDefaultSwiftTest.mailsShouldBeWellReceived » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithDefaultSwiftTest.oneHundredMailsShouldBeWellReceived » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithEncryptedAwsS3Test.mailsContentWithUnicodeCharactersShouldBeKeptUnChanged » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithEncryptedAwsS3Test.mailsShouldBeWellReceived » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithEncryptedAwsS3Test.oneHundredMailsShouldBeWellReceived » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithEncryptedSwiftTest.mailsContentWithUnicodeCharactersShouldBeKeptUnChanged » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithEncryptedSwiftTest.mailsShouldBeWellReceived » IllegalState
[90c76879491515e439067b44d60e15296c6ba42d] [ERROR]   WithEncryptedSwiftTest.oneHundredMailsShouldBeWellReceived » IllegalState
```

**Comment by chibenwa** [comment]
Force pushed to solve conflict

**Comment by mbaechler** [comment]
Could you cleanup this PR to make it readable:
* fix the description
* squash fixups
* have NO_REVIEW commits in the start

**Comment by chibenwa** [comment]
Waiting a green build first.

Which is hard to get without vpn access to the ci and such a mounster refactoring.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
[626991fbe64ff48cef9e97189661167968b03cc0] [INFO] Apache James :: Server :: Web Admin server integration tests :: Distributed FAILURE [18:49 min]

[626991fbe64ff48cef9e97189661167968b03cc0] [ERROR] Errors: 
[626991fbe64ff48cef9e97189661167968b03cc0] [ERROR]   RabbitMQEventDeadLettersIntegrationTest.failedDispatchingShouldBeRedeliveredToAllListeners:626->waitForFailedDispatching:647 » ConditionTimeout
```

**Comment by chibenwa** [comment]
```
[626991fbe64ff48cef9e97189661167968b03cc0] [INFO] Apache James :: Server :: Web Admin server integration tests :: Distributed FAILURE [18:49 min]

[626991fbe64ff48cef9e97189661167968b03cc0] [ERROR] Errors: 
[626991fbe64ff48cef9e97189661167968b03cc0] [ERROR]   RabbitMQEventDeadLettersIntegrationTest.failedDispatchingShouldBeRedeliveredToAllListeners:626->waitForFailedDispatching:647 » ConditionTimeout
```

**Comment by chibenwa** [comment]
```
[c963715cad55f63a8359400965f03ca1101a6796] [ERROR] Failed to execute goal on project james-server-task-api: Could not resolve dependencies for project org.apache.james:james-server-task-api:jar:3.6.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus.mojo:mojo-parent:pom:34 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/mojo/mojo-parent/34/mojo-parent-34.pom.part (No such file or directory) -> [Help 1]
```

test this please

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/MailboxListenersLoaderImpl.java [hunk 2]
```diff
@@ -58,16 +61,24 @@ public void configure(HierarchicalConfiguration<ImmutableNode> configuration) {
     public void configure(ListenersConfiguration listenersConfiguration) {
         LOGGER.info("Loading user registered mailbox listeners");
 
-        guiceDefinedListeners.forEach(eventBus::register);
+        ImmutableMap<Group, MailboxListener.GroupMailboxListener> guiceListenersAsMap = guiceDefinedListeners.stream()
+            .collect(Guavate.toImmutableMap(MailboxListener.GroupMailboxListener::getDefaultGroup));
 
-        listenersConfiguration.getListenersConfiguration().stream()
+        ImmutableMap<Group, MailboxListener> registeredListenersAsMap = listenersConfiguration.getListenersConfiguration().stream()
```
Comment (mbaechler) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/MailboxListenersLoaderImpl.java]: what about `userDefinedListeners`?
BTW, I don't see the value of suffixing variable names with their types (i.e. asMap suffix)

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java [hunk 3]
```diff
@@ -172,6 +173,13 @@ default FinalStage quotaManager(QuotaManager quotaManager) {
         }
 
         interface FinalStage {
+            @FunctionalInterface
+            interface InstanciateListener {
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java:178]: s/Instanciate/Instantiate ?

Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java:178]: What it this type's intent?

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/event/ComputeMessageFastViewProjectionListenerTest.java [hunk 4]
```diff
@@ -249,36 +239,6 @@ void shouldKeepPreviewWhenCopyingMessage() throws Exception {
             .isEqualTo(PRECOMPUTED_PROPERTIES_PREVIEW);
     }
 
-    @Test
-    void shouldStoreEventInDeadLettersWhenComputeFastViewPrecomputedPropertiesException() throws Exception {
```
Comment (mbaechler) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/event/ComputeMessageFastViewProjectionListenerTest.java:253]: I can't see how these tests become irrelevant, could you explain?

File: examples/custom-listeners/src/main/java/org/apache/james/examples/custom/listeners/SetCustomFlagOnBigMessages.java [hunk 5]
```diff
@@ -58,6 +60,7 @@
 
     @Inject
     SetCustomFlagOnBigMessages(MailboxManager mailboxManager) {
+        Preconditions.checkNotNull(mailboxManager);
```
Comment (mbaechler) [inline examples/custom-listeners/src/main/java/org/apache/james/examples/custom/listeners/SetCustomFlagOnBigMessages.java:63]: is it tested anywhere?

File: mailbox/api/src/main/java/org/apache/james/mailbox/events/GroupsAlreadyRegistered.java [hunk 6]
```diff
@@ -19,15 +19,6 @@
 
 package org.apache.james.mailbox.events;
 
-public class GroupAlreadyRegistered extends RuntimeException {
-    private final Group group;
+public class GroupsAlreadyRegistered extends RuntimeException {
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/events/GroupsAlreadyRegistered.java:22]: please suffix the type with `Exception` (yes, I know, it's painful but I think it's a coding standard we always respect)

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 7]
```diff
@@ -758,6 +759,8 @@ void deleteMailboxByIdShouldFireMailboxDeletionEvent() throws Exception {
 
         @Test
         void createMailboxShouldFireMailboxAddedEvent() throws Exception {
+            groupListener.reset(); //needed as setUp did lead to events being dispatched
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java:762]: is it because we register the EventCollector earlier? What about calling `.reset` in `setUp` then?

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/InVMEventBus.java [hunk 8]
```diff
@@ -53,6 +55,7 @@ public InVMEventBus(EventDelivery eventDelivery, RetryBackoffConfiguration retry
         this.eventDeadLetters = eventDeadLetters;
         this.registrations = Multimaps.synchronizedSetMultimap(HashMultimap.create());
         this.groups = new ConcurrentHashMap<>();
```
Comment (mbaechler) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/InVMEventBus.java:57]: You probably lost a fixup but you probably don't need a ConcurrentHashMap because you can only write once to this field.

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/InVMEventBus.java [hunk 9]
```diff
@@ -53,6 +55,7 @@ public InVMEventBus(EventDelivery eventDelivery, RetryBackoffConfiguration retry
...
+        this.initialized = false;
```
Comment (mbaechler) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/InVMEventBus.java:58]: and the boolean is probably useless

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RabbitMQEventBus.java [hunk 10]
```diff
@@ -127,9 +130,13 @@ public Registration register(MailboxListener listener, RegistrationKey key) {
     }
 
     @Override
-    public Registration register(MailboxListener listener, Group group) {
+    public void initialize(Map<Group, MailboxListener> listeners) throws GroupsAlreadyRegistered {
```
Comment (mbaechler) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/RabbitMQEventBus.java]: you don't protect against concurrency as for InVM?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/MemoryMailboxManagerProvider.java [hunk 11]
```diff
@@ -42,4 +43,19 @@ public static InMemoryMailboxManager provideMailboxManager(Set<PreDeletionHook>
             .getMailboxManager();
     }
 
+    public static InMemoryMailboxManager provideMailboxManager(Set<PreDeletionHook> preDeletionHooks, MailboxListener.GroupMailboxListener listener) {
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/MemoryMailboxManagerProvider.java:46]: could we avoid this duplication?

File: mailbox/api/src/main/java/org/apache/james/mailbox/events/EventBusSupplier.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import javax.inject.Provider;
+
+import com.google.common.base.Preconditions;
+
+public class EventBusSupplier implements UninitializedEventBus, Provider<EventBus> {
+    private final UninitializedEventBus uninitializedEventBus;
+    private Optional<EventBus> eventBus;
+
+    @Inject
+    public EventBusSupplier(UninitializedEventBus uninitializedEventBus) {
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/events/EventBusSupplier.java:35]: could you either put the constructor private or annotate as visible for testing?

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 13]
```diff
@@ -120,10 +120,10 @@
 
     protected abstract T provideMailboxManager();
 
-    protected abstract EventBus retrieveEventBus(T mailboxManager);
-
     protected abstract T provideMailboxManager(MailboxListener.GroupMailboxListener mailboxListener);
 
+    protected abstract EventBus retrieveEventBus(T mailboxManager);
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java:125]: why this change?

File: mailbox/api/src/main/java/org/apache/james/mailbox/events/EventBusSupplier.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public EventBus initialize(Map<Group, MailboxListener> listeners) throws GroupsAlreadyRegistered {
+        EventBus eventBus = uninitializedEventBus.initialize(listeners);
+        this.eventBus = Optional.of(eventBus);
+        return eventBus;
+    }
+
+    @Override
+    public EventBus get() {
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/events/EventBusSupplier.java:47]: this method should not exist: it defeats the strong typing strategy.
We should register the Provider in guice to have compile-time check of initialization

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java (hunk 3 above)
Comment (chibenwa) [reply mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/manager/InMemoryIntegrationResources.java:178]: This interface allow to instantiate a MailboxListener from main mailbox components

File: mailbox/api/src/main/java/org/apache/james/mailbox/events/EventBusSupplier.java (hunk 14 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/events/EventBusSupplier.java:47]: I don't care about guice I care about mailbox unit tests, mpt tests tests.

**Comment by chibenwa** [comment]
@mbaechler let's forget that API seggregation thing it is clearly going nowhere given your review comments.

How about merging https://github.com/linagora/james-project/pull/3279 instead?

**Comment by chibenwa** [comment]
No longer relevant.

Read https://github.com/linagora/james-project/pull/3323
